internal/app/machined/pkg/controllers/block/lvm.go

```diff
@@ -0,0 +1,132 @@
+type LVMActivationController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *LVMActivationController) Name() string {
+	return "block.LVMActivationController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *LVMActivationController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: v1alpha1.NamespaceName,
+			Type:      runtimeres.MountStatusType,
+			ID:        optional.Some(constants.EphemeralPartitionLabel),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: block.NamespaceName,
+			Type:      block.DiscoveredVolumeType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *LVMActivationController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *LVMActivationController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	touchedIDs := make(map[string]struct{})
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+			if _, err := safe.ReaderGetByID[*runtimeres.MountStatus](ctx, r, constants.EphemeralPartitionLabel); err != nil {
+				if state.IsNotFoundError(err) {
+					// wait for the mount status to be available
+					continue
+				}
+
+				return fmt.Errorf("failed to get mount status: %w", err)
+			}
+		}
+
+		discoveredVolumes, err := safe.ReaderListAll[*block.DiscoveredVolume](ctx, r)
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("failed to list discovered volumes: %w", err)
+		}
+
+		for iterator := discoveredVolumes.Iterator(); iterator.Next(); {
+			if iterator.Value().TypedSpec().Name != "lvm2-pv" {
+				continue
+			}
+
+			// first we check if all associated volumes are available
+			// https://man7.org/linux/man-pages/man7/lvmautoactivation.7.html
+			stdOut, err := cmd.RunContext(ctx,
+				"/sbin/lvm",
+				"pvscan",
+				"--cache",
+				"--verbose",
+				"--listvg",
+				"--checkcomplete",
+				"--vgonline",
+				"--autoactivation",
+				"event",
+				"--udevoutput",
+				iterator.Value().TypedSpec().DevPath,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to check LVM volume availability: %w", err)
+			}
+
+			if stdOut == "" || !strings.HasPrefix(stdOut, "LVM_VG_NAME_COMPLETE") {
+				continue
+			}
+
+			vgName := strings.TrimSuffix(strings.TrimPrefix(strings.TrimSuffix(stdOut, "\n"), "LVM_VG_NAME_COMPLETE='"), "'")
+
+			if _, ok := touchedIDs[vgName]; ok {
+				continue
+			}
+
+			logger.Info("activating LVM volume", zap.String("name", vgName))
+
+			// activate the volume group
+			if _, err = cmd.RunContext(ctx,
+				"/sbin/lvm",
+				"vgchange",
+				"-aay",
+				"--autoactivation",
+				"event",
+				vgName,
+			); err != nil {
+				return fmt.Errorf("failed to activate LVM volume %s: %w", vgName, err)
+			}
+
+			touchedIDs[vgName] = struct{}{}
```

Comment:
we could move `touchedIDs` out to the `type LVMActivationController struct{}`, so that it survives controller restarts

Reply:
the udev rule was always getting triggered, but the command output is empty and it skips that way, i wonder if we need `touchedIDs` at all :thinking:

Reply:
I think it had something like "first time seen" in the list of steps as well.